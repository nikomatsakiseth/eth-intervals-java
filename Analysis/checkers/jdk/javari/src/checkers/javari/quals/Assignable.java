package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that a field is assignable, even if it is inside a {@link ReadOnly}
 * instance.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Assignable {

}
