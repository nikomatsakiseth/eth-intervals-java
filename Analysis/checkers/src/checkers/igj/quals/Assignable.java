package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.*;

/**
 *
 * Indicates the annotated {@code Field} may be re-assigned regardless of the
 * immutability of the enclosing class or object instance.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target( { FIELD } )
public @interface Assignable {

}
