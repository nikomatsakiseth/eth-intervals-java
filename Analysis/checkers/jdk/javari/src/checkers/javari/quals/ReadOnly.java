package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import static java.lang.annotation.ElementType.*;

/**
 * Indicates that, for the variable on which this annotation appears,
 * the object to which this variable refers will not be modified via
 * this reference, except its fields explicitly marked as
 * {@link Mutable}.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see Mutable
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target({FIELD, LOCAL_VARIABLE, METHOD, PARAMETER, TYPE})
@TypeQualifier
@SubtypeOf({})
public @interface ReadOnly {

}
