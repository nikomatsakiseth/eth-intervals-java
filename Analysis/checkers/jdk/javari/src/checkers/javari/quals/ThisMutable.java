package checkers.javari.quals;

import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * An annotation used to represent a place holder immutability type, that is
 * equivalent to the ThisMutable type on the Javari typesystem.
 *
 * However, it is an implementation detail; hence, the package-scope.
 */
@TypeQualifier
@Target({}) // A programmer cannot write @ThisMutable in a program
@SubtypeOf(ReadOnly.class)
public @interface ThisMutable {

}
