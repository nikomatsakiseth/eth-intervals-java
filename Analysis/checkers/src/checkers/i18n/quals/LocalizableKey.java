package checkers.i18n.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import checkers.quals.Unqualified;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface LocalizableKey { }
