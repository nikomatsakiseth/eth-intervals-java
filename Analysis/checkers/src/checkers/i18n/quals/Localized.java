package checkers.i18n.quals;

import static com.sun.source.tree.Tree.Kind.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.*;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@ImplicitFor( trees = {
        /* All integer literals */
        INT_LITERAL,
        LONG_LITERAL,
        FLOAT_LITERAL,
        DOUBLE_LITERAL,
        BOOLEAN_LITERAL,

        //CHAR_LITERAL,
        //STRING_LITERAL,
        //NULL_LITERAL
        }
)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface Localized { }
