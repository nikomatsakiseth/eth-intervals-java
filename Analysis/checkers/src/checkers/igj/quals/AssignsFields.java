package checkers.igj.quals;

import static java.lang.annotation.ElementType.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 *
 * Indicates that the annotated method could assigns (but not mutate) the fields
 * of {@code this} object.
 *
 * A method with an AssignsFields receiver may not use the receiver to
 * invoke other methods with mutable or immutable reciever.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
// TODO: Document this

@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { CONSTRUCTOR, METHOD })
@TypeQualifier // (for now)
@SubtypeOf( ReadOnly.class )
public @interface AssignsFields {

}
