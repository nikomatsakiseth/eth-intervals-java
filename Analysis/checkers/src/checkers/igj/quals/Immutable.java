package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.*;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that the annotated reference is an immutable reference to an
 * immutable object.
 *
 * An Immutable object cannot be modified. Its fields may be reassigned or
 * mutated only if they are explicitly marked as Mutable or Assignable.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { FIELD, LOCAL_VARIABLE, CONSTRUCTOR, METHOD, PARAMETER, TYPE })
@TypeQualifier
@SubtypeOf( ReadOnly.class )
public @interface Immutable {

}
