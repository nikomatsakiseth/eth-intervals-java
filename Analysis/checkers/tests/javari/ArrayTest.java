import checkers.javari.quals.*;

class ArrayTest {         int @ReadOnly [] rmi;

    @Mutable Object aMM @Mutable[];
    @Mutable Object aMTm[];
    @ReadOnly Object aMRo @Mutable [];
    @Mutable Object aTmM[];
    Object aTmTm[];
    @ReadOnly Object aTmRo[];
    @Mutable Object aRoM @ReadOnly [];
    Object aRoTm @ReadOnly [];
    @ReadOnly Object aRoRo @ReadOnly [];

    void testPrimitiveArray() {
        int a;
        @ReadOnly int b;               // error
        int[] mmi;
        int @ReadOnly [] rmi;
        @ReadOnly int[] mri;            // error
        @ReadOnly int @ReadOnly [] rri;  // error
        int[][] mmmi;
        int[] @ReadOnly [] rmmi;
        int @ReadOnly [][] mrmi;
        int @ReadOnly [] @ReadOnly [] rrmi;
        @ReadOnly int[][] mmri;        // error
        @ReadOnly int[] @ReadOnly [] rmri; // error
        @ReadOnly int @ReadOnly [][] mrri;  // error
        @ReadOnly int @ReadOnly [] @ReadOnly [] rrri; // error

    }

    void testAM() {

        // aMM = ...
        aMM = aMM;
        aMM = aMTm;
        aMM = aMRo;    // error
        aMM = aTmM;
        aMM = aTmTm;
        aMM = aTmRo;   // error
        aMM = aRoM;    // error
        aMM = aRoTm;   // error
        aMM = aRoRo;   // error

        // aMTm = ...
        aMTm = aMM;
        aMTm = aMTm;
        aMTm = aMRo;   // error
        aMTm = aTmM;
        aMTm = aTmTm;
        aMTm = aTmRo;  // error
        aMTm = aRoM;   // error
        aMTm = aRoTm;  // error
        aMTm = aRoRo;  // error

        // aMRo = ...
        aMRo = aMM;
        aMRo = aMTm;
        aMRo = aMRo;
        aMRo = aTmM;
        aMRo = aTmTm;
        aMRo = aTmRo;
        aMRo = aRoM;   // error
        aMRo = aRoTm;  // error
        aMRo = aRoRo;  // error

    }

    void testATm() {

        // aTmM = ...
        aTmM = aMM;
        aTmM = aMTm;
        aTmM = aMRo;   // error
        aTmM = aTmM;
        aTmM = aTmTm;
        aTmM = aTmRo;  // error
        aTmM = aRoM;   // error
        aTmM = aRoTm;  // error
        aTmM = aRoRo;  // error

        // aTmTm = ...
        aTmTm = aMM;
        aTmTm = aMTm;
        aTmTm = aMRo;  // error
        aTmTm = aTmM;
        aTmTm = aTmTm;
        aTmTm = aTmRo; // error
        aTmTm = aRoM;  // error
        aTmTm = aRoTm; // error
        aTmTm = aRoRo; // error

        // aTmRo = ...
        aTmRo = aMM;
        aTmRo = aMTm;
        aTmRo = aMRo;
        aTmRo = aTmM;
        aTmRo = aTmTm;
        aTmRo = aTmRo;
        aTmRo = aRoM;  // error
        aTmRo = aRoTm; // error
        aTmRo = aRoRo; // error

    }

    void testARo() {

        // aRoM = ...
        aRoM = aMM;
        aRoM = aMTm;
        aRoM = aMRo;   // error
        aRoM = aTmM;
        aRoM = aTmTm;
        aRoM = aTmRo;  // error
        aRoM = aRoM;
        aRoM = aRoTm;
        aRoM = aRoRo;  // error

        // aRoTm = ...
        aRoTm = aMM;
        aRoTm = aMTm;
        aRoTm = aMRo;  // error
        aRoTm = aTmM;
        aRoTm = aTmTm;
        aRoTm = aTmRo; // error
        aRoTm = aRoM;
        aRoTm = aRoTm;
        aRoTm = aRoRo; // error

        // aRoRo = ...
        aRoRo = aMM;
        aRoRo = aMTm;
        aRoRo = aMRo;
        aRoRo = aTmM;
        aRoRo = aTmTm;
        aRoRo = aTmRo;
        aRoRo = aRoM;
        aRoRo = aRoTm;
        aRoRo = aRoRo;

    }

    void testAMZero() {

        // aMM = ...
        aMM[0] = aMM[0];
        aMM[0] = aMTm[0];
        aMM[0] = aMRo[0];    // error
        aMM[0] = aTmM[0];
        aMM[0] = aTmTm[0];
        aMM[0] = aTmRo[0];   // error
        aMM[0] = aRoM[0];
        aMM[0] = aRoTm[0];
        aMM[0] = aRoRo[0];   // error

        // aMTm = ...
        aMTm[0] = aMM[0];
        aMTm[0] = aMTm[0];
        aMTm[0] = aMRo[0];   // error
        aMTm[0] = aTmM[0];
        aMTm[0] = aTmTm[0];
        aMTm[0] = aTmRo[0];  // error
        aMTm[0] = aRoM[0];
        aMTm[0] = aRoTm[0];
        aMTm[0] = aRoRo[0];  // error

        // aMRo = ...
        aMRo[0] = aMM[0];
        aMRo[0] = aMTm[0];
        aMRo[0] = aMRo[0];
        aMRo[0] = aTmM[0];
        aMRo[0] = aTmTm[0];
        aMRo[0] = aTmRo[0];
        aMRo[0] = aRoM[0];
        aMRo[0] = aRoTm[0];
        aMRo[0] = aRoRo[0];

    }

    void testATmZero() {

        // aTmM = ...
        aTmM[0] = aMM[0];
        aTmM[0] = aMTm[0];
        aTmM[0] = aMRo[0];   // error
        aTmM[0] = aTmM[0];
        aTmM[0] = aTmTm[0];
        aTmM[0] = aTmRo[0];  // error
        aTmM[0] = aRoM[0];
        aTmM[0] = aRoTm[0];
        aTmM[0] = aRoRo[0];  // error

        // aTmTm = ...
        aTmTm[0] = aMM[0];
        aTmTm[0] = aMTm[0];
        aTmTm[0] = aMRo[0];  // error
        aTmTm[0] = aTmM[0];
        aTmTm[0] = aTmTm[0];
        aTmTm[0] = aTmRo[0]; // error
        aTmTm[0] = aRoM[0];
        aTmTm[0] = aRoTm[0];
        aTmTm[0] = aRoRo[0]; // error

        // aTmRo = ...
        aTmRo[0] = aMM[0];
        aTmRo[0] = aMTm[0];
        aTmRo[0] = aMRo[0];
        aTmRo[0] = aTmM[0];
        aTmRo[0] = aTmTm[0];
        aTmRo[0] = aTmRo[0];
        aTmRo[0] = aRoM[0];
        aTmRo[0] = aRoTm[0];
        aTmRo[0] = aRoRo[0];

    }

    void testARoZero() {

        // aRoM = ...
        aRoM[0] = aMM[0];
        aRoM[0] = aMTm[0];
        aRoM[0] = aMRo[0];   // error
        aRoM[0] = aTmM[0];
        aRoM[0] = aTmTm[0];
        aRoM[0] = aTmRo[0];  // error
        aRoM[0] = aRoM[0];
        aRoM[0] = aRoTm[0];
        aRoM[0] = aRoRo[0];  // error

        // aRoTm = ...
        aRoTm[0] = aMM[0];
        aRoTm[0] = aMTm[0];
        aRoTm[0] = aMRo[0];  // error
        aRoTm[0] = aTmM[0];
        aRoTm[0] = aTmTm[0];
        aRoTm[0] = aTmRo[0]; // error
        aRoTm[0] = aRoM[0];
        aRoTm[0] = aRoTm[0];
        aRoTm[0] = aRoRo[0]; // error

        // aRoRo = ...
        aRoRo[0] = aMM[0];
        aRoRo[0] = aMTm[0];
        aRoRo[0] = aMRo[0];
        aRoRo[0] = aTmM[0];
        aRoRo[0] = aTmTm[0];
        aRoRo[0] = aTmRo[0];
        aRoRo[0] = aRoM[0];
        aRoRo[0] = aRoTm[0];
        aRoRo[0] = aRoRo[0];

    }
}
