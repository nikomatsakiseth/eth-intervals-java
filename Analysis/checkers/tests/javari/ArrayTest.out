:17: (primitive.ro)
:20: (primitive.ro)
:21: (primitive.ro)
:26: (primitive.ro)
:27: (primitive.ro)
:28: (primitive.ro)
:29: (primitive.ro)
:38: (type.incompatible)
:41: (type.incompatible)
:42: (type.incompatible)
:43: (type.incompatible)
:44: (type.incompatible)
:49: (type.incompatible)
:52: (type.incompatible)
:53: (type.incompatible)
:54: (type.incompatible)
:55: (type.incompatible)
:64: (type.incompatible)
:65: (type.incompatible)
:66: (type.incompatible)
:75: (type.incompatible)
:78: (type.incompatible)
:79: (type.incompatible)
:80: (type.incompatible)
:81: (type.incompatible)
:86: (type.incompatible)
:89: (type.incompatible)
:90: (type.incompatible)
:91: (type.incompatible)
:92: (type.incompatible)
:101: (type.incompatible)
:102: (type.incompatible)
:103: (type.incompatible)
:112: (type.incompatible)
:115: (type.incompatible)
:118: (type.incompatible)
:123: (type.incompatible)
:126: (type.incompatible)
:129: (type.incompatible)
:149: (type.incompatible)
:152: (type.incompatible)
:155: (type.incompatible)
:160: (type.incompatible)
:163: (type.incompatible)
:166: (type.incompatible)
:186: (type.incompatible)
:189: (type.incompatible)
:192: (type.incompatible)
:197: (type.incompatible)
:200: (type.incompatible)
:203: (type.incompatible)
:223: (type.incompatible)
:226: (type.incompatible)
:229: (type.incompatible)
:234: (type.incompatible)
:237: (type.incompatible)
:240: (type.incompatible)
